Update neomerx/jsonapi to ^3.0.0

dakota opened the pull request:
This closes #57 and upgrades crud-json-api to use the latest version of neomerx/jsonapi. This also changes the api to use (JSON:API v1.1)[https://jsonapi.org/format/1.1]

dakota committed:
Start support for neomerx/json-api 3.x

dakota committed:
Update dynamic schema to use neomarx/jsonapi 3.0 interfaces

dakota committed:
Update error renderer

dakota committed:
Update view class

dakota committed:
Fix issue with ResultSetDecorator

dakota committed:
Fix includes

dakota committed:
Trigger afterFind for afterPaginate

dakota committed:
WIP: Update tests

dakota committed:
Update integration tests

dakota committed:
Update all tests

dakota committed:
Fix PHPCS errors

dakota committed:
Fix PHPStan errors

dakota committed:
Was not supposed to do those

dakota committed:
Include a lowest deps test

dakota committed:
Minimum CakePHP 3.7

bravo-kernel commented:
Wow, this is amazing, thank you so much. Do you need me to merge this once it is done or do you have merge permission yourself?

dakota commented:
I have merge permission :)

I think it is done, all tests pass, and my own application seems to be working with it. I would appreciate a 2nd pair of eyes looking at the code before it gets merged if you've got the time.

bravo-kernel commented:
Fair enough, I will try to look at it later today and report back here.

dakota committed:
Small fix to allow multi-key links for N:1 or 1:1

dakota committed:
Use values for self route

dakota committed:
Correct PHPCS errors

dakota committed:
Base link controller name on actual table class name, instead of registry alias

dakota committed:
Update tests

bravo-kernel commented on tests/TestCase/View/JsonApiViewTest.php:
```diff
@@ -281,18 +287,18 @@ public function testEncodeWithoutSchemas()
-        ];
-
-        $this->assertArraySubset($expectedVersionArray, json_decode($view->render(), true));
+//        $countries = TableRegistry::get('Countries')->find()->all();
```
Should this not be removed?

bravo-kernel commented:
Besides the commented-out code it all looks good to me so feel free to merge and create a new release.

dakota committed:
Automatically include belongsTo as relationships

dakota committed:
Use a shallow toArray method for entities to prevent child entities being converted twice

dakota committed:
Infer the correct plugin when generating links

dakota committed:
De-dasherize sort fields

dakota committed:
If pagination sort isn't set, but the sort query is, use the sort query

dakota committed:
Adds a searchListener for advanced searches.

Advanced JSON:API searches require using a nested `filter` query param. With the search behaviour, that means using `filter.date` type query params. However, that makes it tricky to share search collections/functionality between JSON:API endpoints and traditional endpoints. Using this custom behaviour just for JSON:API requests takes care of that.

dakota committed:
Add an empty array as default to prevent breakage when no filters are passed

dakota committed:
Work to get parcel saving to work

dakota committed:
JSON:API type parameters are now more predictable

dakota committed:
Correct JSON:Api version number in exception

bravo-kernel commented:
👏

dakota committed:
Allow for including extra top level links from the controller

dakota committed:
Update fixture files

dakota committed:
Update all tests

dakota committed:
Fix PHPCS and PHPStan errors

dakota committed:
Fix incorrect change

bravo-kernel commented:
Awesome job mr. @dakota 👏
